flink-connector-snowflake/src/main/java/io/deltastream/flink/connector/snowflake/sink/internal/SnowflakeSinkServiceImpl.java

```diff
@@ -274,47 +293,24 @@ private void handleInsertRowsErrors(List<InsertValidationResponse.InsertError> e
                 errors.get(0).getException());
     }
 
-    /**
-     * Reopen the ingestion channel on {@link SFException} using {@link dev.failsafe.Fallback}.
-     *
-     * @param row {@link java.util.Map} to send to Snowflake for eventual flush
-     * @return {@link InsertValidationResponse}
-     */
-    private InsertValidationResponse insertRowWithFallback(final Map<String, Object> row) {
-        Fallback<Object> reopenChannelFallbackExecutorForInsertRows =
-                Fallback.builder(
-                                attempt -> {
-                                    this.ingestionFallbackSupplier(attempt.getLastException());
-                                })
-                        .handle(SFException.class)
-                        .onFailedAttempt(
-                                event ->
-                                        LOGGER.warn(
-                                                "Failed to send row to ingest channel",
-                                                event.getLastException()))
-                        .onFailure(
-                                event ->
-                                        LOGGER.error(
-                                                String.format(
-                                                        "[INSERT_ROW_FALLBACK] Failed to re-open channel '%s'",
-                                                        this.getChannelName()),
-                                                event.getException()))
-                        .build();
-
-        return Failsafe.with(reopenChannelFallbackExecutorForInsertRows)
-                .get(() -> this.channel.insertRow(row, null));
-    }
-
-    /**
-     * Fail on any errors that may have happened during ingestion to the Snowflake service.
-     *
-     * @param throwable {@link Throwable}
-     */
-    private void ingestionFallbackSupplier(final Throwable throwable) {
-        LOGGER.warn(
-                "[INSERT_ROWS_FALLBACK] Failed to insert row with channel '{}'. Exiting with error",
-                this.getChannelName());
-        throw new RuntimeException(throwable);
+    protected long getLatestCommittedOffsetFromSnowflakeIngestChannel() {
+        Map<String, String> offsetTokens =
+                this.getClient().getLatestCommittedOffsetTokens(List.of(this.getChannel()));
+        Preconditions.checkState(
+                offsetTokens.size() == 1,
+                String.format(
+                        "Expected getLatestCommittedOffsetTokens to return information for a single channel. Found %s offset tokens.",
+                        offsetTokens.size()));
+        String offsetToken = offsetTokens.get(offsetTokens.keySet().iterator().next());
+        try {
+            return ObjectUtils.isEmpty(offsetToken) ? 0 : Long.parseLong(offsetToken);
```

Comment:
This can strictly check for String:
```suggestion
            return StringUtils.isEmpty(offsetToken) ? 0 : Long.parseLong(offsetToken);
```